Source/DafnyCore/AST/Expressions/Expressions.cs

```diff
@@ -783,6 +783,26 @@ public abstract class Expression : TokenNode {
     return e;
   }
 
+  /// <summary>
+  /// For a given collection type "collectionType", return the ResolvedOpcode and appropriate BoundedPool
+  /// used in creating a resolved expression of the form "x in collection", where "collection" is given as
+  /// a parameter.
+  /// </summary>
+  public static void ComputeResolvedOpcodeAndBoundedPool(Expression collection, CollectionType collectionType,
```

Comment:
I would still change this signature to `CollectionBoundedPool CollectionType.ComputeBoundedPool(Expression collection)`, and add a property `ResolveOpCode CollectionBoundedPool.OpCode`. If you add a new CollectionType, don't you want the implementor to think about how to create a BoundedPool for it?

In general, I think up to and including the resolution phase, so that includes parsing printing and resolution, all code specific to an AST type should live inside its definition.

Reply:
Okay, I created `CollectionType.GetBoundedPool` as you suggested. I also created a `ResolvedOpcodeForIn` property, but I place it in `CollectionType` as well (rather than in `CollectionBoundedPool`), since it could be useful outside the context of bounded pools, too.